lib/semantics/canonicalize-do.cc

```diff
@@ -0,0 +1,118 @@
+      if constexpr (std::is_same_v<T, common::Indirection<parser::EndDoStmt>>) {
+        std::get<parser::ExecutableConstruct>(currentIter_->u).u =
+            parser::Statement<parser::ActionStmt>{
+                std::optional<parser::Label>{currentLabel},
+                parser::ContinueStmt{}};
+      }
+      do {
+        currentIter_ = MakeCanonicalForm(labelDoIters_.back(), currentIter_);
+        labelDoIters_.pop_back();
+        labels_.pop_back();
+      } while (!labels_.empty() && labels_.back() == currentLabel);
+    }
+    return false;
+  }
+
+private:
+  parser::Block SpliceBlock(
+      parser::Block::iterator beginLoop, parser::Block::iterator endLoop) {
+    parser::Block block;
+    block.splice(block.begin(), *currentList_, ++beginLoop, ++endLoop);
+    return block;
+  }
+  std::optional<parser::LoopControl> CreateLoopControl(
+      std::optional<parser::LoopControl> &loopControlOpt) {
+    if (loopControlOpt.has_value()) {
+      return std::optional<parser::LoopControl>(
+          parser::LoopControl{loopControlOpt->u});
+    }
+    return std::optional<parser::LoopControl>{};
+  }
+  std::optional<parser::LoopControl> ExtractLoopControl(
+      const parser::Block::iterator &startLoop) {
+    return CreateLoopControl(std::get<std::optional<parser::LoopControl>>(
+        std::get<parser::Statement<common::Indirection<parser::LabelDoStmt>>>(
+            std::get<parser::ExecutableConstruct>(startLoop->u).u)
+            .statement->t));
+  }
+  parser::Block::iterator MakeCanonicalForm(
+      const parser::Block::iterator &startLoop,
+      const parser::Block::iterator &endLoop) {
+    std::get<parser::ExecutableConstruct>(startLoop->u).u =
+        common::Indirection<parser::DoConstruct>{std::make_tuple(
+            parser::Statement<parser::NonLabelDoStmt>{
+                std::optional<parser::Label>{},
+                parser::NonLabelDoStmt{
+                    std::make_tuple(std::optional<parser::Name>{},
+                        ExtractLoopControl(startLoop))}},
+            SpliceBlock(startLoop, endLoop),
+            parser::Statement<parser::EndDoStmt>{std::optional<parser::Label>{},
+                parser::EndDoStmt{std::optional<parser::Name>{}}})};
+    return startLoop;
+  }
+  parser::Block::iterator CheckStatement(const parser::Block::iterator &iter) {
+    currentIter_ = iter;
+    parser::ExecutionPartConstruct &executionPartConstruct{*iter};
+    if (auto *executableConstruct = std::get_if<parser::ExecutableConstruct>(
+            &executionPartConstruct.u)) {
+      if (auto *labelDoLoop = std::get_if<
+              parser::Statement<common::Indirection<parser::LabelDoStmt>>>(
+              &executableConstruct->u)) {
+        labelDoIters_.push_back(iter);
+        labels_.push_back(std::get<parser::Label>(labelDoLoop->statement->t));
+      } else if (!labels_.empty()) {
+        parser::Walk(executableConstruct->u, *this);
+      }
+    }
+    return currentIter_;
```

Comment:
`currentIter_` is mysterious hidden shared state.  I don't know what it points at or which member functions can change it.

Reply:
I'll add a comment.